arch/src/aarch64/fdt.rs

```diff
@@ -547,107 +542,138 @@ fn create_devices_node<T: DeviceInfoForFdt + Clone + Debug, S: ::std::hash::Buil
-        // device addresses
-        0x3000000,                            // (ss = 11: 64-bit memory space)
-        (pci_device_base_64bit >> 32) as u32, // PCI address
-        pci_device_base_64bit as u32,
-        (pci_device_base_64bit >> 32) as u32, // CPU address
-        pci_device_base_64bit as u32,
-        (pci_device_size_64bit >> 32) as u32, // size
-        pci_device_size_64bit as u32,
-    ];
-    let bus_range = [0, 0]; // Only bus 0
-    let reg = [PCI_MMCONFIG_START.0, PCI_MMCONFIG_SIZE];
-
-    let pci_node = fdt.begin_node("pci")?;
-    fdt.property_string("compatible", "pci-host-ecam-generic")?;
-    fdt.property_string("device_type", "pci")?;
-    fdt.property_array_u32("ranges", &ranges)?;
-    fdt.property_array_u32("bus-range", &bus_range)?;
-    fdt.property_u32("#address-cells", 3)?;
-    fdt.property_u32("#size-cells", 2)?;
-    fdt.property_array_u64("reg", &reg)?;
-    fdt.property_u32("#interrupt-cells", 1)?;
-    fdt.property_null("interrupt-map")?;
-    fdt.property_null("interrupt-map-mask")?;
-    fdt.property_null("dma-coherent")?;
-    fdt.property_u32("msi-parent", MSI_PHANDLE)?;
-
-    if let Some(virtio_iommu_bdf) = virtio_iommu_bdf {
-        // See kernel document Documentation/devicetree/bindings/pci/pci-iommu.txt
-        // for 'iommu-map' attribute setting.
-        let iommu_map = [
+    // In multiple PCI segments setup, each PCI segment needs a PCI node.
+    for pci_device_info_elem in pci_device_info.iter() {
+        // EDK2 requires the PCIe high space above 4G address.
+        // The actual space in CLH follows the RAM. If the RAM space is small, the PCIe high space
+        // could fall bellow 4G.
+        // Here we put it above 4G in FDT to workaround the EDK2 check.
+        // But the address written in ACPI is not impacted.
+        let pci_device_base_64bit: u64 = if cfg!(feature = "acpi") {
+            pci_device_info_elem.pci_device_space_start + PCI_HIGH_BASE
+        } else {
+            pci_device_info_elem.pci_device_space_start
+        };
+        let pci_device_size_64bit: u64 = if cfg!(feature = "acpi") {
+            pci_device_info_elem.pci_device_space_size - PCI_HIGH_BASE
+        } else {
+            pci_device_info_elem.pci_device_space_size
+        };
+        // 4K boundary (?)
+        let pci_device_size_32bit: u64 =
+            MEM_32BIT_DEVICES_SIZE / ((1 << 12) * pci_device_info.len() as u64) * (1 << 12);
+        let pci_device_base_32bit: u64 = MEM_32BIT_DEVICES_START.0
+            + pci_device_size_32bit * pci_device_info_elem.pci_segment_id as u64;
+
+        let ranges = [
+            // io addresses
+            0x1000000,
             0_u32,
-            VIRTIO_IOMMU_PHANDLE,
             0_u32,
-            virtio_iommu_bdf,
-            virtio_iommu_bdf + 1,
-            VIRTIO_IOMMU_PHANDLE,
-            virtio_iommu_bdf + 1,
-            0xffff - virtio_iommu_bdf,
+            (MEM_PCI_IO_START.0 >> 32) as u32,
```

Comment:
Setting the same base and size for all segments is logically incorrect. But as I remember, this setting is only used to comfort EDK2 checking. The setting doesn't take any effect for AArch64.

So, do you mind adding some comments here to explain?

Reply:
IO space is not used on arm64, so we can set this part exactly the same for each segment. I added more comments to explain this.